### PR TITLE
KEYCLOAK-16363 Add scope mappings to realm/client CRDs

### PR DESCRIPTION
**Pull request by dhx-teuto**
## JIRA ID
https://issues.redhat.com/browse/KEYCLOAK-16363

This pull request is only intended for the last commit, the others are minor fixes to code generation etc. that should be done beforehand.
I will rebase as needed.

## Checklist:
- [x] Automated Tests
- [ ] [Documentation](https://github.com/keycloak/keycloak-documentation) changes if necessary
- [x] (If required) Discussed on [Keycloak DEV](https://groups.google.com/forum/#!forum/keycloak-dev)

### CODE REVIEW

**Comment by slaskawi** [comment]
I'll hold off with comments until we get your previous PRs in. Then, I'll ask you for a rebase and we'll take it from there.

**Comment by slaskawi** [comment]
@dhx-teuto Could you please rebase this? I'd like to get it in shortly.

File: test/e2e/keycloak_client_test.go [hunk 1]
```diff
@@ -192,3 +223,145 @@ func waitForClientRoles(t *testing.T, framework *test.Framework, keycloakCR keyc
...
+		retrievedMappings.RealmMappings,
+		expected.RealmMappings)
+	assert.Equal(t, 0, len(difference))
+	assert.Equal(t, 2, len(intersection))
+
+	difference, intersection = keycloakclient.RoleDifferenceIntersection(
+		retrievedMappings.ClientMappings[secondClientName].Mappings,
+		expected.ClientMappings[secondClientName].Mappings)
+	assert.Equal(t, 0, len(difference))
+	assert.Equal(t, 2, len(intersection))
```
Comment (dhx-teuto) [inline test/e2e/keycloak_client_test.go:364]: @slaskawi I changed the visibility of `RoleDifferenceIntersection`to reuse it in the test.

**Comment by dhx-teuto** [comment]
> @dhx-teuto Could you please rebase this? I'd like to get it in shortly.

@slaskawi Did you have a chance to look at this yet?

File: pkg/apis/keycloak/v1alpha1/keycloakrealm_types.go [hunk 2]
```diff
@@ -164,6 +164,13 @@ type KeycloakAPIRealm struct {
 	// Roles
 	// +optional
 	Roles *RolesRepresentation `json:"roles,omitempty"`
+
+	// Scope Mappings
+	// +optional
+	ScopeMappings []ScopeMappingRepresentation `json:"scopeMappings,omitempty"`
+	// Client Scope Mappings
```
Comment (slaskawi) [inline pkg/apis/keycloak/v1alpha1/keycloakrealm_types.go:171]: nitpick: new line

File: pkg/common/client_state.go [hunk 3]
```diff
@@ -59,6 +60,15 @@ func (i *ClientState) Read(context context.Context, cr *kc.KeycloakClient, realm
...
 		}
+
+		i.ScopeMappings, err = realmClient.ListScopeMappings(cr.Spec.Client.ID, i.Realm.Spec.Realm.Realm)
+		if err != nil {
+			return err
+		}
+	}
+
+	if i.ScopeMappings == nil {
+		i.ScopeMappings = &kc.MappingsRepresentation{}
```
Comment (slaskawi) [inline pkg/common/client_state.go:71]: In most of the cases `ClientState` uses nils (that's why it uses pointers). Is there any reason to break this convention here and use an empty object?

File: pkg/controller/keycloakclient/keycloakclient_reconciler.go [hunk 4]
```diff
@@ -122,6 +124,68 @@ func roleMatches(a kc.RoleRepresentation, b kc.RoleRepresentation) bool {
...
+	mappingsDeleted := scopeMappingDifference(state.ScopeMappings, cr.Spec.ScopeMappings)
+	if mappingsDeleted.RealmMappings != nil {
+		desired.AddAction(i.getDeletedClientRealmScopeMappingsState(state, cr, &mappingsDeleted.RealmMappings))
+	}
+	for _, clientMappings := range mappingsDeleted.ClientMappings {
+		desired.AddAction(i.getDeletedClientClientScopeMappingsState(state, cr, clientMappings.DeepCopy()))
+	}
+}
+
+func scopeMappingDifference(a *kc.MappingsRepresentation, b *kc.MappingsRepresentation) (d *kc.MappingsRepresentation) {
```
Comment (slaskawi) [inline pkg/controller/keycloakclient/keycloakclient_reconciler.go:152]: The logic in this function is not very straightforward. Some function comment would be useful here.

File: test/e2e/keycloak_client_test.go (hunk 1 above)
Comment (slaskawi) [reply test/e2e/keycloak_client_test.go:364]: Honestly, I don't like this. It make a weird dependency between the `e2e` package and the reconciler. A dependency that should never be there. In the `model` we have a `util.go` - maybe `RoleDifferenceIntersection` should go there?

**Review by slaskawi — CHANGES_REQUESTED** [review]

File: pkg/apis/keycloak/v1alpha1/keycloakrealm_types.go (hunk 2 above)
Comment (dhx-teuto) [reply pkg/apis/keycloak/v1alpha1/keycloakrealm_types.go:171]: I consider those two a group, if you want me to change it anyway let me know.

File: pkg/common/client_state.go (hunk 3 above)
Comment (dhx-teuto) [reply pkg/common/client_state.go]: This was an oversight, I just deleted it, thanks.

File: pkg/controller/keycloakclient/keycloakclient_reconciler.go (hunk 4 above)
Comment (dhx-teuto) [reply pkg/controller/keycloakclient/keycloakclient_reconciler.go:126]: done

File: test/e2e/keycloak_client_test.go (hunk 1 above)
Comment (dhx-teuto) [reply test/e2e/keycloak_client_test.go]: good idea, done

**Comment by dhx-teuto** [comment]
@slaskawi ready for review again

File: pkg/controller/keycloakclient/keycloakclient_reconciler.go (hunk 4 above)
Comment (slaskawi) [reply pkg/controller/keycloakclient/keycloakclient_reconciler.go:126]: Thanks!

**Review by slaskawi — APPROVED** [review]

**Comment by slaskawi** [comment]
Integrated, thanks @dhx-teuto !

**Pull request merged**